### PR TITLE
P2585R1 Improving default container formatting

### LINKED ISSUE #5624: [2022-07 LWG Motion 37] P2585R1 Improving default container formatting
[P2585R1](https://wg21.link/p2585r1)

Must be applied after [2022-07 LWG Motion 14] P2286R8 Formatting Ranges, #5601

### PR DESCRIPTION
**Pull request by burblebee**
Fixes #5624 

Based on P2286R8 Formatting Ranges in branch motions-2022-07-lwg-14

### CODE REVIEW

**Comment by JohelEGP** [comment]
Shouldn't this be merged into said branch, rather than `main`? Like other motion PRs that modify existing motions.

File: source/utilities.tex [hunk 1]
```diff
@@ -15581,6 +15644,215 @@
...
+
+\indexlibraryglobal{\exposid{range-default-formatter}}%
+\begin{codeblock}
+namespace std {
+  template<ranges::@\libconcept{input_range}@ R, class charT>
+  struct @\exposid{range-default-formatter}@<range_format::set, R, charT> {
+  private:
+    using @\exposid{maybe-const-set}@ = @\exposid{fmt-maybe-const}@<R, charT>;                          // \expos
+    range_formatter<remove_cvref_t<ranges::range_reference_t<maybe-const-set>>,
+                    charT> @\exposid{underlying_}@;                                         // \expos
```
Comment (JohelEGP) [inline source/utilities.tex:15732]:
```suggestion
    using @\exposid{maybe-const-set}@ = @\exposid{fmt-maybe-const}@<R, charT>;                            // \expos
    range_formatter<remove_cvref_t<ranges::range_reference_t<@\exposid{maybe-const-set}@>>,   // \expos
                    charT> @\exposid{underlying_}@;
```

File: source/utilities.tex [hunk 2]
```diff
@@ -15581,6 +15644,215 @@
...
+namespace std {
+  template<ranges::@\libconcept{input_range}@ R, class charT>
+  struct @\exposid{range-default-formatter}@<range_format::map, R, charT> {
+  private:
+    using @\exposid{maybe-const-map}@ = @\exposid{fmt-maybe-const}@<R, charT>;                        // \expos
+    using @\exposid{element-type}@ =                                                      // \expos
+      remove_cvref_t<ranges::range_reference_t<@\exposid{maybe-const-map}@>>;
+    range_formatter<@\exposid{element-type}@, charT> @\exposid{underlying_}@;                         // \expos
+  public:
+    constexpr formatter();
```
Comment (JohelEGP) [inline source/utilities.tex:15658]:
```suggestion
    constexpr @\exposid{range-default-formatter}@();
```

File: source/utilities.tex [hunk 3]
```diff
@@ -15581,6 +15644,215 @@
...
+    template<class FormatContext>
+      typename FormatContext::iterator
+        format(@\exposid{maybe-const-map}@& r, FormatContext& ctx) const;
+  };
+}
+\end{codeblock}
+
+\indexlibrarymember{formatter}{\exposid{range-default-formatter}}%
+\begin{itemdecl}
+constexpr formatter();
```
Comment (JohelEGP) [inline source/utilities.tex:15673]:
```suggestion
constexpr @\exposid{range-default-formatter}@();
```

File: source/utilities.tex [hunk 4]
```diff
@@ -15581,6 +15644,215 @@
...
+namespace std {
+  template<ranges::@\libconcept{input_range}@ R, class charT>
+  struct @\exposid{range-default-formatter}@<range_format::set, R, charT> {
+  private:
+    using @\exposid{maybe-const-set}@ = @\exposid{fmt-maybe-const}@<R, charT>;                          // \expos
+    range_formatter<remove_cvref_t<ranges::range_reference_t<maybe-const-set>>,
+                    charT> @\exposid{underlying_}@;                                         // \expos
+
+  public:
+    constexpr formatter();
```
Comment (JohelEGP) [inline source/utilities.tex:15735]:
```suggestion
    constexpr @\exposid{range-default-formatter}@();
```

File: source/utilities.tex [hunk 5]
```diff
@@ -15581,6 +15644,215 @@
...
+      typename FormatContext::iterator
+        format(@\exposid{maybe-const-set}@& r, FormatContext& ctx) const;
+  };
+}
+\end{codeblock}
+
+%% FIXME: what's the return type? No longer a ctor.
+\indexlibrarymember{formatter}{\exposid{range-default-formatter}}%
+\begin{itemdecl}
+constexpr formatter();
```
Comment (JohelEGP) [inline source/utilities.tex:15751]:
```suggestion
\indexlibrarymember{formatter}{\exposid{range-default-formatter}}%
\begin{itemdecl}
constexpr @\exposid{range-default-formatter}@();
```
Good catch. I hadn't noticed. It'd have been easy to catch if the edits and move weren't mingled in the same commit. The editor should be able to deal with those commits just like with fixup commits.

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/utilities.tex (hunk 5 above)
Comment (jwakely) [reply source/utilities.tex:15751]: Yeah, obviously the answer to the question is "there's no return type, it's still supposed to be a ctor". The intention was not to change it from a ctor to a normal member function with a missing return type. The ctor should just have been renamed along with its enclosing class, but LWG missed that.

File: source/utilities.tex [hunk 6]
```diff
@@ -15051,6 +15283,579 @@
...
+\begin{itemdescr}
+\pnum
+The type of \tcode{r} is \tcode{const R\&}
+if \tcode{\libconcept{input_range}<const R>} is \tcode{true} and
+\tcode{R\&} otherwise.
+
+\pnum
+\effects
+Let \tcode{\placeholder{s}} be a \tcode{basic_string<charT>} such that
+\tcode{ranges::equal(s, r)} is \tcode{true}.
```
Comment (JohelEGP) [inline source/utilities.tex:15855]:
```suggestion
\tcode{ranges::equal(\placeholder{s}, r)} is \tcode{true}.
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/utilities.tex (hunk 1 above)
Comment (burblebee) [reply source/utilities.tex:15735]: Do we really want this change? The // \expos is currently on the same line as the identifier it applies to, which seems preferable to me

Comment (JohelEGP) [reply source/utilities.tex:15735]: Yikes! Sorry, I thought the identifier was on the previous line.

File: source/utilities.tex (hunk 2 above)
Comment (burblebee) [reply source/utilities.tex:15661]: Thanks, that clarifies my main issue with the paper :). done

File: source/utilities.tex (hunk 5 above)
Comment (burblebee) [reply source/utilities.tex:15754]: Thanks guys, fixed

File: source/utilities.tex (hunk 6 above)
Comment (burblebee) [reply source/utilities.tex:15855]: fixed

**Comment by burblebee** [comment]
> Shouldn't this be merged into said branch, rather than `main`? Like other motion PRs that modify existing motions.

Hopefully it's clear how to apply this.  The edits were rebased on top of motions-2022-07-lwg-14, so once motions-2022-07-lwg-14 is merged, this pr can be rebased and merged.  There should only be two commits to apply:

* P2585R1 Improving default container formatting [258cfea]
* [format.range.formatter] Rename constructor to match renamed struct [a2bee27]

**Comment by JohelEGP** [comment]
Also resolves https://github.com/cplusplus/papers/issues/1244.

**Pull request merged**